lib/puppetserver/ca/sign_action.rb

```diff
@@ -0,0 +1,179 @@
+require 'puppetserver/ca/utils'
+require 'puppetserver/utils/http_utilities'
+require 'puppetserver/utils/file_utilities'
+require 'puppetserver/ca/puppet_config'
+require 'optparse'
+require 'openssl'
+require 'net/https'
+require 'json'
+
+module Puppetserver
+  module Ca
+    class SignAction
+
+      include Puppetserver::Utils
+
+      SUMMARY = 'Sign a given certificate'
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca sign [--help|--version]
+  puppetserver ca sign [--config] --certname CERTNAME[,CERTNAME]
+  puppetserver ca sign  --all
+
+Description:
+Given a comma-separated list of valid certnames, instructs the CA to sign each cert.
+
+Options:
+      BANNER
+      BODY = JSON.dump({desired_state: 'signed'})
+
+      def self.parser(parsed = {})
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--certname x,y,z', Array, 'the name(s) of the cert(s) to be signed') do |cert|
+            parsed['certname'] = cert
+          end
+          opts.on('--config PUPPET.CONF', 'Custom path to Puppet\'s config file') do |conf|
+            parsed['config'] = conf
+          end
+          opts.on('--help', 'Display this command specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+          opts.on('--all', 'Operate on all certnames') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run(input)
+        config = input['config']
+
+        if config
+          errors = FileUtilities.validate_file_paths(config)
+          return 1 if Utils.handle_errors(@logger, errors)
+        end
+
+        puppet = PuppetConfig.parse(config)
+        return 1 if Utils.handle_errors(@logger, puppet.errors)
+
+        if input['all']
+          requested_certnames = get_all_pending_certs(puppet.settings)
+          if requested_certnames.nil?
+            return 1
+          end
+        else
+          requested_certnames = input['certname']
+        end
+
+        success = sign_requested_certs(requested_certnames, puppet.settings)
+        return success ? 0 : 1
+      end
+
+      def get_certificate_statuses(settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server], settings[:ca_port],
+                                        'certificate_statuses', 'any_key')
+        HttpUtilities.with_connection(url, settings) do |connection|
+          connection.get(url)
+        end
+      end
+
+      def sign_cert(certnames,settings)
+        results = { }
+        url = HttpUtilities.make_ca_url(settings[:ca_server], settings[:ca_port],
+                                        'certificate_status')
+        HttpUtilities.with_connection(url, settings) do |connection|
```

Comment:
@amysahli Is this better, do you think? I needed to retain the certnames as well as the results (because we use both for error logging) so I just made it pass back a hash.

Reply:
yeah, i thinks that looks good